narwhals/functions.py

```diff
@@ -88,15 +96,60 @@ def from_dict(data: dict[str, Any], *, native_namespace: Any) -> DataFrame[Any]:
         └─────┴─────┘
     """
     if native_namespace is get_polars():
-        native_frame = native_namespace.from_dict(data)
+        if schema:
+            from narwhals._polars.utils import (
+                reverse_translate_dtype as polars_reverse_translate_dtype,
+            )
+
+            schema = {
+                name: polars_reverse_translate_dtype(dtype)
+                for name, dtype in schema.items()
+            }
+
+        native_frame = native_namespace.from_dict(data, schema=schema)
     elif (
         native_namespace is get_cudf()
         or native_namespace is get_modin()
         or native_namespace is get_pandas()
     ):
         native_frame = native_namespace.DataFrame.from_dict(data)
+
+        if schema:
+            from narwhals._pandas_like.utils import Implementation
+            from narwhals._pandas_like.utils import (
+                reverse_translate_dtype as pandas_like_reverse_translate_dtype,
+            )
+
+            implementation = (
+                Implementation.PANDAS
+                if native_namespace is get_pandas()
+                else Implementation.MODIN
+                if native_namespace is get_modin()
+                else Implementation.CUDF
+            )
```

Comment:
I may have a tiny refactor in mind for this

Reply:
Nice, I like that!

Maybe we should even expose `Implementation`, and `from_dict` could take `implementation` as an argument?

We haven't made a release yet so we can still make this change 😎

Reply:
Let me open the PR to refactor implementation in the first place then 😇